dumb_pypi/main.py

```diff
@@ -196,18 +209,50 @@ def _format_datetime(dt: datetime) -> str:
     return dt.strftime('%Y-%m-%d %H:%M:%S')
 
 
-def _package_json(files: List[Package], base_url: str) -> Dict[str, Any]:
+IMPORTANT_METADATA_FOR_INFO = frozenset((
+    'name',
+    'version',
+    'requires_dist',
+    'requires_python',
+))
+
+
+def _package_json(sorted_files: List[Package], base_url: str) -> Dict[str, Any]:
     # https://warehouse.pypa.io/api-reference/json.html
     # note: the full api contains much more, we only output the info we have
-    by_version: Dict[str, List[Dict[str, Any]]] = collections.defaultdict(list)
-    for file in files:
+    by_version: Dict[str, List[Package]] = collections.defaultdict(list)
+    for file in sorted_files:
         if file.version is not None:
-            by_version[file.version].append(file.json_info(base_url))
+            by_version[file.version].append(file)
+
+    # Find a file from the latest release to use for "info". We don't want to
+    # mix-and-match the metadata across releases since tools like Poetry rely
+    # on this, but we do want to pick the file in the release with the most
+    # populated metadata.
+    latest_file = sorted_files[-1]
+    if sorted_files[-1].version is not None:
+        latest_file = max(
+            by_version[sorted_files[-1].version],
+            key=lambda f: sum(bool(getattr(f, v)) for v in IMPORTANT_METADATA_FOR_INFO),
+        )
 
     return {
-        'info': {'name': files[0].name, 'version': files[0].version},
-        'releases': by_version,
-        'urls': by_version[files[0].version] if files[0].version else [],
```

Comment:
The old logic here was actually picking the oldest file to feature in `info` and `urls` since the list is sorted oldest-first, whoops.